js/modules/k6/experimental/tracing/tracing.go

```diff
@@ -0,0 +1,286 @@
+package tracing
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/dop251/goja"
+	"go.k6.io/k6/js/common"
+	"go.k6.io/k6/js/modules"
+	"go.k6.io/k6/metrics"
+)
+
+// Tracing is the JS module instance that will be created for each VU.
+type Tracing struct {
+	vu modules.VU
+
+	propagator Propagator
+}
+
+// InstrumentHTTP instruments the HTTP module with tracing headers.
+//
+// When used in the context of a k6 script, it will automatically replace
+// the imported http module's methods with instrumented ones.
+func (t *Tracing) InstrumentHTTP(options instrumentationOptions) {
+	if err := t.configure(options); err != nil {
+		common.Throw(t.vu.Runtime(), err)
+	}
+
+	// Explicitly inject the http module in the VU's runtime.
+	// This allows us to later on override the http module's methods
+	// with instrumented ones.
+	httpModuleValue, err := t.vu.Runtime().RunString(`require('k6/http')`)
+	if err != nil {
+		common.Throw(t.vu.Runtime(), err)
+	}
+
+	httpModuleObj := httpModuleValue.ToObject(t.vu.Runtime())
+
+	HTTPMethods := [...]k6HTTPMethodName{
+		k6HTTPDeleteMethodName,
+		k6HTTPGetMethodName,
+		k6HTTPHeadMethodName,
+		k6HTTPOptionsMethodName,
+		k6HTTPPatchMethodName,
+		k6HTTPPostMethodName,
+		k6HTTPPutMethodName,
+		k6HTTPRequestMethodName,
+	}
+
+	for _, method := range HTTPMethods {
+		originalMethodFn, ok := goja.AssertFunction(httpModuleObj.Get(string(method)))
+		if !ok {
+			common.Throw(t.vu.Runtime(), fmt.Errorf("http.%s is not a function", method))
+		}
+
+		tracedMethodFn := t.instrumentHTTPMethod(method, originalMethodFn)
+
+		// Inject the new get function, adding tracing headers
+		// to the request in the HTTP module object.
+		err = httpModuleObj.Set(string(method), tracedMethodFn)
+		if err != nil {
+			common.Throw(t.vu.Runtime(), err)
+		}
+	}
+
+	// Inject the updated HTTP module object in the runtime,
+	// overriding any previously imported one in the process.
+	err = t.vu.Runtime().Set("http", httpModuleObj)
+	if err != nil {
+		common.Throw(t.vu.Runtime(), err)
+	}
```

Comment:
I have 0 knowledge of how k6 actually works internally, but anytime I see something like this, a 💡 appears in my head. Can this be overridden by another k6 module that overrides HTTP? I understand that this module does not wrap the HTTP module serving as a proxy but rather replaces it with a modified HTTP client. If another module does the same thing, tracing will suddenly stop working, correct?

Reply:
I see your concern, but I don't believe that would be the case (there might be some corner case or JS devilry escaping me, though).

The tracing module imports the `http` module, and wraps its method to insert its logic before and after the execution of the underlying `http.{method}`. The call to the HTTP method itself is left untouched; all we do in the case of the tracing module is essentially modify the params argument the user gave to the call to the HTTP method so that it includes the tracing context, set metadata, calls the HTTP method with the updated args, delete the metadata, and return the result of the method. Thus, even if another method used the same technique, it would still work as long as they don't modify the function's prototype.